pkg/nodeagent/controller/operatingsystemconfig/reconciler.go

```diff
@@ -0,0 +1,318 @@
+		return fmt.Errorf("unable to create temporary directory: %w", err)
+	}
+
+	for _, file := range files {
+		if file.Content.Inline == nil {
+			continue
+		}
+
+		if err := r.FS.MkdirAll(filepath.Dir(file.Path), fs.ModeDir); err != nil {
+			return fmt.Errorf("unable to create directory %q: %w", file.Path, err)
+		}
+
+		permissions := fs.FileMode(0600)
+		if file.Permissions != nil {
+			permissions = fs.FileMode(*file.Permissions)
+		}
+
+		data, err := extensionsv1alpha1helper.Decode(file.Content.Inline.Encoding, []byte(file.Content.Inline.Data))
+		if err != nil {
+			return fmt.Errorf("unable to decode data of file %q: %w", file.Path, err)
+		}
+
+		tmpFilePath := filepath.Join(tmpDir, filepath.Base(file.Path))
+		if err := afero.WriteFile(r.FS, tmpFilePath, data, permissions); err != nil {
+			return fmt.Errorf("unable to create temporary file %q: %w", tmpFilePath, err)
+		}
+
+		if err := r.FS.Rename(tmpFilePath, file.Path); err != nil {
+			return fmt.Errorf("unable to rename temporary file %q to %q: %w", tmpFilePath, file.Path, err)
+		}
+
+		log.Info("Successfully applied new or changed file", "path", file.Path)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) removeDeletedFiles(log logr.Logger, files []extensionsv1alpha1.File) error {
+	for _, file := range files {
+		if err := r.FS.Remove(file.Path); err != nil && !errors.Is(err, afero.ErrFileNotFound) {
+			return fmt.Errorf("unable to delete no longer needed file %q: %w", file.Path, err)
+		}
+
+		log.Info("Successfully removed no longer needed file", "path", file.Path)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) applyChangedUnits(ctx context.Context, log logr.Logger, units []changedUnit) error {
+	for _, unit := range units {
+		unitFilePath := path.Join(etcdSystemdSystem, unit.Name)
+
+		if unit.Content != nil {
+			oldUnitContent, err := afero.ReadFile(r.FS, unitFilePath)
+			if err != nil && !errors.Is(err, afero.ErrFileNotFound) {
+				return fmt.Errorf("unable to read existing unit file %q for %q: %w", unitFilePath, unit.Name, err)
+			}
+
+			newUnitContent := []byte(*unit.Content)
+			if !bytes.Equal(newUnitContent, oldUnitContent) {
+				if err := afero.WriteFile(r.FS, unitFilePath, newUnitContent, 0600); err != nil {
+					return fmt.Errorf("unable to write unit file %q for %q: %w", unitFilePath, unit.Name, err)
+				}
+				log.Info("Successfully applied new or changed unit file", "path", unitFilePath)
+			}
+		}
+
+		for _, dropIn := range unit.DropIns {
+			dropInFilePath := path.Join(unitFilePath+".d", dropIn.Name)
+
+			oldDropInContent, err := afero.ReadFile(r.FS, dropInFilePath)
+			if err != nil && !errors.Is(err, afero.ErrFileNotFound) {
+				return fmt.Errorf("unable to read existing drop-in file %q for unit %q: %w", dropInFilePath, unit.Name, err)
+			}
+
+			newDropInContent := []byte(dropIn.Content)
+			if !bytes.Equal(newDropInContent, oldDropInContent) {
+				if err := afero.WriteFile(r.FS, dropInFilePath, newDropInContent, 0600); err != nil {
```

Comment:
I don't understand, can you explain, please? The unit files/drop-in files are always created with `0600` permissions and this is not configurable. Hence, it will never change unless we change this code here. Why do we need the `Chmod` then?

Reply:
I was thinking about restoring the original permissions in case someone changed them manually on the node.

Reply:
> I was thinking about restoring the original permissions in case someone changed them manually on the node.

Whut ? This should never happen and should be treated as malicious

Reply:
Though, it could happen (users have full permissions and can do whatever they want on the node's file system). I think explicitly setting the file permissions is okay as part of this reconciler. Will do 👍🏻